pandas/core/internals/managers.py

```diff
@@ -384,8 +384,29 @@ def setitem(self, indexer, value) -> Self:
             raise ValueError(f"Cannot set values with ndim > {self.ndim}")
 
         if using_copy_on_write() and not self._has_no_reference(0):
-            # if being referenced -> perform Copy-on-Write and clear the reference
-            # this method is only called if there is a single block -> hardcoded 0
+            # Split blocks to only copy the columns we want to modify, only 1 block
+            if self.ndim == 2 and isinstance(indexer, tuple):
+                blk_loc = self.blklocs[indexer[1]]
+                if is_list_like(blk_loc) and blk_loc.ndim == 2:
```

Comment:
how would you get blk_loc.ndim == 2?

Reply:
Would have to check, got test failures without checking this

Reply:
Did you check this again?

Reply:
Forget to comment ``maybe_convert_indexer`` makes ``indexer[1]`` 2D in some cases which causes ``blk_loc.ndim=2``

Reply:
What indexer[1] causes this?  this strikes me as very weird

Reply:
test_loc_setitem_reordering_with_all_true_indexer

That's one of the test that hits this if you want to take a look